include/libp2p/protocol/gossip/impl/gossip_wire_protocol.hpp

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef LIBP2P_GOSSIP_WIRE_PROTOCOL_HPP
+#define LIBP2P_GOSSIP_WIRE_PROTOCOL_HPP
+
+#include <libp2p/protocol/gossip/common.hpp>
+
+#include <map>
+
+#include <gsl/span>
+
+namespace pubsub::pb {
+  class RPC;
+  class ControlMessage;
+}  // namespace pubsub::pb
+
+namespace libp2p::protocol::gossip {
+
+  /// Subscribe-unsubscribe request
+  struct SubOpts {
+    /// true means subscribe, false means unsubscribe
+    bool subscribe = false;
+
+    /// Pub-sub topic
+    TopicId topic_id;
+  };
+
+  /// Announces about topic messages available on a host
+  using IHaveTable = std::map<TopicId, Repeated<MessageId>>;
+
+  /// General wire protocol message
+  struct RPCMessage {
+    /// Subscribe-unsubscribe requests
+    Repeated<SubOpts> subscriptions;
+
+    /// "I have" announces by topics
+    IHaveTable i_have;
+
+    /// Requests for messages
+    Repeated<MessageId> i_want;
+
+    /// Requests to join per-topic mesh
+    Repeated<TopicId> graft;
+
+    /// Requests to leave per-topic mesh
+    Repeated<TopicId> prune;
+
+    /// Messages to publish
+    Repeated<TopicMessage::Ptr> publish;
+
+    bool empty() const;
+
+    void clear();
+
+    bool deserialize(gsl::span<const uint8_t> bytes);
+
+    bool serialize(std::vector<uint8_t> &buffer) const;
+  };
+
+  /// Interface for accepting sub messages being read from wire
+  class MessageReceiver {
+   public:
+    virtual ~MessageReceiver() = default;
+
+    /// Topic subscribe-unsubscribe request received
+    virtual void onSubscription(const PeerId &from, bool subscribe,
+                                const TopicId &topic) = 0;
+
+    /// "I have message ids" notification received
+    virtual void onIHave(const PeerId &from, const TopicId &topic,
+                         MessageId &&msg_id) = 0;
+
+    /// "I want message" request received
+    virtual void onIWant(const PeerId &from, MessageId &&msg_id) = 0;
+
+    /// Graft request received (gossip mesh control)
+    virtual void onGraft(const PeerId &from, const TopicId &topic) = 0;
+
+    /// Prune request received (gossip mesh control)
+    virtual void onPrune(const PeerId &from, const TopicId &topic) = 0;
+
+    /// Message received
+    virtual void onMessage(const PeerId &from, TopicMessage::Ptr msg) = 0;
+  };
+
+  /// Parses RPC protobuf message received from wire
+  bool parseRPCMessage(const PeerId &from, gsl::span<const uint8_t> bytes,
```

Comment:
This function does two things: parses message and sends parsed message to receiver. 
Split this logic into two or more functions.